homeassistant/components/calendar/trigger.py

```diff
@@ -41,34 +42,135 @@
 # mypy: disallow-any-generics
 
 
+@dataclass
+class QueuedCalendarEvent:
+    """An event that is queued to be fired in the future."""
+
+    trigger_time: datetime.datetime
+    event: CalendarEvent
+
+
+@dataclass
+class Timespan:
+    """A time range part of start/end dates, used for considering active events."""
+
+    start: datetime.datetime
+    """The start datetime of the interval."""
+
+    end: datetime.datetime
+    """The end datetime (exclusive) of the interval."""
+
+    def with_offset(self, offset: datetime.timedelta) -> Timespan:
+        """Return a new interval shifted by the specified offset."""
+        return Timespan(self.start + offset, self.end + offset)
+
+    def contains(self, trigger: datetime.datetime) -> bool:
```

Comment:
You could go all fancy and do `__in__ = contains` and I think it would allow to do `if timestamp in timespan:`